test/cmocka/src/math/matrix/ref_matrix.m

```diff
@@ -0,0 +1,186 @@
+	opt.b_qxy_x = 1;
+	opt.b_qxy_y = 15;
+	opt.c_qxy_x = 1;
+	opt.c_qxy_y = 15;
+	opt.elementwise = 1;
+	get_ref_mult(opt);
+
+	opt.test_n = 4;
+	opt.input = 'random';
+	opt.a_rows = 5;
+	opt.a_columns = 1;
+	opt.b_rows = 1;
+	opt.b_columns = 4;
+	opt.a_qxy_x = 1;
+	opt.a_qxy_y = 15;
+	opt.b_qxy_x = 1;
+	opt.b_qxy_y = 15;
+	opt.c_qxy_x = 1;
+	opt.c_qxy_y = 15;
+	opt.elementwise = 0;
+	get_ref_mult(opt);
+end
+
+function get_ref_mult(opt)
+	[a_bits, a_scale] = get_qxy_scale(opt.a_qxy_x, opt.a_qxy_y);
+	[b_bits, b_scale] = get_qxy_scale(opt.b_qxy_x, opt.b_qxy_y);
+	[c_bits, c_scale] = get_qxy_scale(opt.c_qxy_x, opt.c_qxy_y);
+	bits = max([a_bits b_bits c_bits]);
+	if (a_bits ~= b_bits)
+		error('Matrix word lengths need to be the same (for now)');
+	end
+
+	% Data to process
+	switch lower(opt.input)
+		case 'ones'
+			a = ones(opt.a_rows, opt.a_columns);
+			b = ones(opt.b_rows, opt.b_columns);
+		case 'random'
+			a = (rand(opt.a_rows, opt.a_columns) * 2 - 1) * a_scale;
+			b = (rand(opt.b_rows, opt.b_columns) * 2 - 1) * b_scale;
+		otherwise
+			error('Illegal input');
+	end
+
+	[iqa, fqa] = quant_qxy(a, opt.a_qxy_x, opt.a_qxy_y);
+	[iqb, fqb] = quant_qxy(b, opt.b_qxy_x, opt.b_qxy_y);
+	if opt.elementwise
+		c = fqa .* fqb;
+	else
+		c = fqa * fqb;
+	end
+
+	iqc = quant_qxy(c, opt.c_qxy_x, opt.c_qxy_y);
+	c_size = size(c);
+
+	header_fn = sprintf('ref_matrix_mult_%d_test%d.h', bits, opt.test_n);
+	define_prefix = sprintf('MATRIX_MULT_%d_TEST%d_', bits, opt.test_n);
+	vector_prefix = sprintf('matrix_mult_%d_test%d_', bits, opt.test_n);
+
+	fh = export_headerfile_open(header_fn);
+	comment = sprintf('This file was created %s with script ref_matrix.m', ...
+			  datestr(now, 0));
+	export_comment(fh, comment);
+	export_ndefine(fh, [define_prefix 'ELEMENTWISE'], opt.elementwise);
+	export_ndefine(fh, [define_prefix 'A_ROWS'], opt.a_rows);
+	export_ndefine(fh, [define_prefix 'A_COLUMNS'], opt.a_columns);
+	export_ndefine(fh, [define_prefix 'B_ROWS'], opt.b_rows);
+	export_ndefine(fh, [define_prefix 'B_COLUMNS'], opt.b_columns);
+	export_ndefine(fh, [define_prefix 'C_ROWS'], c_size(1));
+	export_ndefine(fh, [define_prefix 'C_COLUMNS'], c_size(2));
+	export_ndefine(fh, [define_prefix 'A_QXY_Y'], opt.a_qxy_y);
+	export_ndefine(fh, [define_prefix 'B_QXY_Y'], opt.b_qxy_y);
+	export_ndefine(fh, [define_prefix 'C_QXY_Y'], opt.c_qxy_y);
+
+	export_vector(fh, 'int16_t', [vector_prefix 'a'], mat_to_vec(iqa));
+	export_vector(fh, 'int16_t', [vector_prefix 'b'], mat_to_vec(iqb));
+	export_vector(fh, 'int16_t', [vector_prefix 'c'], mat_to_vec(iqc));
+	fprintf(1, 'Exported file %s\n', header_fn);
+end
+
+function [bits, scale] = get_qxy_scale(qxy_x, qxy_y)
+	bits = qxy_x + qxy_y;
+	scale = (2^(bits - 1) - 1) / 2^qxy_y;
+end
+
+function v = mat_to_vec(m)
+	s = size(m);
+	v = zeros(s(1) * s(2), 1);
+	for i = 1:s(1)
+		idx1 = s(2) * (i - 1) + 1;
+		idx2 = idx1 + s(2) - 1;
+		v(idx1:idx2) = m(i, :);
+	end
+end
+
+function [y,qval] = quant_qxy(val, x, y)
```

Comment:
Which rounding approach is desired, i.e. ceiling, floor, value rounded to absolute value, or zero? 
Is it possible to round to the nearest permissible quantized value?

Reply:
Matlab round() is my de-facto round and it's compatible with libc ... Are there benefits from using some other way?